enhancements/ingress/customize-error-code-pages.md

```diff
@@ -0,0 +1,121 @@
+
+### Non-Goal
+TBD
+
+## Proposal
+0. User creates a custom error code page configmap `http503page` in `openshift-config`
+1. The IngressController API is extended by adding an optional
+`HttpErrorCodePage` field with type string
+```go
+//This used to sync the configmap having customizable error code page created by cluster admin in openshift-config namespace.
+//to openshift-ingress which will be used by the router to serve appropriate error code pages.
+//If this field is empty, the default values i.e default shipped error code pages will be served by the haproxy router.
+HttpErrorCodePage string `json:"httpErrorCodePage,omitempty"`
+
+```
+Using the default, the router uses the default error pages used by haproxy.
+The `HttpErrorCodePage` field enables the cluster administrator to
+specify custom error code pages.
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpErrorCodePage: "http503page"  // http503page is the config map name with custom 
+                                    //503 error page created in openshift-config by cluster-admin. 
+```
+2. `errorfile` stanzas need to be added as per https://www.haproxy.com/blog/serve-dynamic-custom-error-pages-with-haproxy/
+3. default error pages will be added in openshift/router for the haproxy.
+4. A  new controller to sync the cluster admin created configmap having custom error pages from `openshift-config` 
+   to `openshift-ingress`.
+5. changes to ingress controller's function `desiredRouterDeployment` present in `deployment.go`.
+   of openshift/cluster-ingress-operator to mount the custom error code page configmap into the router deployment.
+
+
+### Validation
+Omitting `spec.httpErrorCodePage`  specifies the default behavior i.e serves default error code pages.
+If user specifies an invalid value, then the IngressController ignores the annotation.
+
+### User Stories
+
+#### As a cluster administrator, I expect to have custom error code pages getting served
+1. The administrator creates a configmap named http503page in the openshift-config namespace.
+2. The administrator patches the ingresscontroller to reference the http503page configmap by name.
+3. The ingress operator copies the http503page configmap from the openshift-config namespace to the
+   openshift-ingress namespace.
+4. The ingress operator configures the router deployment with a volume and volume mount that mounts
+   the custom error pages from the http503page configmap on top of the existing error pages in the 
+   router pods.
+   
+### Risks and Mitigations
+No risk identified. This is a cosmetic change.
+
+## Design Details
+
+### Test Plan
+Unit test cases will be added for the new controller which sync the configmap from openshift-config
+to openshift-ingress.
+Unit test cases will be added for the ingress controller where the router deployment get a volume mount
+of the custom error code page
```

Comment:
I would suggestion adding some details about a potential e2e test for this enhancement.

Reply:
OK